rfd/0194-upstream-pki-support.md

```diff
@@ -0,0 +1,279 @@
+  operator to ensure that this is not possible or does not pose a risk.
+- Validators MUST take additional care when processing X509 certificates. There
+  is no longer a guarantee that a certificate that passes signature validation is
+  a valid X509 SVID that has been issued in line with the SPIFFE specification.
+- When a trust domain federates with a Teleport Workload Identity trust domain,
+  the federating trust domain will be trusting X509 certificates issued anywhere
+  within the PKI hierarchy rather than solely X509 SVIDs issued by the trust 
+  domain.
+
+Additionally, end-users should be made aware of other implications such as:
+
+- Whilst X509 SVIDs will be trusted across the entire PKI hierarchy, JWT SVIDs
+  will only be trusted within the Teleport Workload Identity trust domain. This
+  creates a disparity between X509 and JWT SVIDs that would not usually exist.
+
+### Phase 2: Automation & Integration
+
+To provide a seamless experience for users, we can offer integrations with
+popular PKI solutions to remove the need to manually complete the CSR process
+each rotation.
+
+There two potential design approaches for this:
+
+- Integration functionality built directly into the Teleport Auth Server.
+- Integrations are built in a similar way to Teleport Access Plugins, small,
+  light-weight binaries that interact with the Teleport API.
+
+Taking an approach similar to the access plugins is likely to be the best choice
+because:
+
+- Decouples the PKI integration from the Teleport Auth Server, allowing us to
+  iterate more quickly.
+- The PKI integration plugin can be run in networks that the Teleport Auth
+  Server cannot access, allowing Teleport Cloud Tenants integrate with a 
+  PKI solution which is on-prem.
+- Ability to publish SDK tooling that allows users to build their own PKI
+  integrations that function in the same way as the first-party integrations.
+
+Each integration will follow a similar pattern:
+
+- The integration is configured with credentials for the Teleport API, and,
+  the credentials necessary to interact with the PKI solution.
+- The integration will poll the configured CA's status via the Teleport API.
+- When the CA approaches a configured threshold of remaining lifetime, the
+  integration will trigger the rotation to the "pre-init" phase.
+- The integration will generate a CSR for the new CA key-pair and submit it to
+  the PKI solution.
+- The integration will poll the PKI solution for the signed certificate and
+  chain.
+- The integration will submit the signed certificate and chain to the Teleport
+  API to complete the `pre-init` phase.
+
+Ideally, the integrations should also ship with a "one-shot" mode which allows
+the user to manually trigger the rotation to the "pre-init" phase and complete
+the rotation in a single step. This could leverage the user's existing `tsh`
+configuration to authenticate with the Teleport API.
```

Comment:
I assume this one-shot rotation is possible because everything trusts the root CA and it doesn't matter if we suddenly switch the intermediate we're using, is that what you're talking about here?